File: python/src/ttio/codecs/base_pack.py
"""BASE_PACK genomic-sequence codec — 2-bit ACGT + sidecar mask.

Clean-room implementation. The 2-bit-per-base packing convention is
decades-old prior art, fundamental and ungatewayed by IP. **No
htslib, no jbzip, no CRAM tools-Java source consulted.** The
sidecar mask layout (sparse position+byte list) is a TTI-O-specific
design choice — see HANDOFF.md M84 §3, binding decision §80.

Cross-language equivalents:
    Objective-C: TTIOBasePack (objc/Source/Codecs/TTIOBasePack.{h,m})
    Java:        global.thalion.ttio.codecs.BasePack

Wire format (big-endian throughout, self-contained):

    Offset      Size  Field
    ──────      ────  ──────────────────────────────────────────
    0           1     version            (0x00)
    1           4     original_length    (uint32 BE)
    5           4     packed_length      (uint32 BE — = ceil(orig/4))
    9           4     mask_count         (uint32 BE)
    13          var   packed_body        (packed_length bytes)
    13+pl       var   mask               (mask_count × 5 bytes:
                                           uint32 BE position,
                                           uint8 original_byte)

Total length = ``13 + packed_length + 5 * mask_count`` bytes.

Pack mapping (case-sensitive; binding decision §81):

    'A' (0x41) → 0b00
    'C' (0x43) → 0b01
    'G' (0x47) → 0b10
    'T' (0x54) → 0b11
    anything else → mask entry (placeholder 0b00 written to body)

Bit order within byte is **big-endian** (binding decision §82) —
the first base in the input occupies the two highest-order bits.
The padding bits in the final body byte (when ``len(input) % 4 !=
0``) are zero (binding decision §83); the decoder uses
``original_length`` to know how many slots to consume.

Mask entries are sorted ascending by position (binding decision
§84); the encoder emits them in input order, the decoder validates
strict ascending order. The first byte is ``version = 0x00``
(binding decision §85), distinct from the M79 codec id ``0x06``.
"""
from __future__ import annotations

import operator
import struct

# ── Wire-format constants ──────────────────────────────────────────

#: Version byte — first byte of every BASE_PACK stream.
VERSION: int = 0x00

#: Header bytes: 1 (version) + 4 (orig_len) + 4 (packed_len) +
#: 4 (mask_count) = 13.
HEADER_LEN: int = 13

#: Mask entry size: uint32 BE position + uint8 original byte.
MASK_ENTRY_LEN: int = 5

# ── Pack lookup tables ─────────────────────────────────────────────

#: 256-entry ``bytes.translate`` table mapping every input byte to a
#: 2-bit slot value: A→0, C→1, G→2, T→3, every other byte → 0
#: (the placeholder written into the body for non-ACGT slots).
def _build_translate_table() -> bytes:
    tbl = bytearray(256)  # default 0 — placeholder for non-ACGT
    tbl[ord("A")] = 0b00
    tbl[ord("C")] = 0b01
    tbl[ord("G")] = 0b10
    tbl[ord("T")] = 0b11
    return bytes(tbl)


#: Companion table for non-ACGT detection: ACGT bytes map to 0x00,
#: every other byte to 0x01. A single ``b"\x01" in marks`` check
#: tells us whether any mask entries are needed.
def _build_mark_table() -> bytes:
    tbl = bytearray(b"\x01" * 256)
    tbl[ord("A")] = 0x00
    tbl[ord("C")] = 0x00
    tbl[ord("G")] = 0x00
    tbl[ord("T")] = 0x00
    return bytes(tbl)


_PACK_TRANSLATE: bytes = _build_translate_table()
_MARK_TRANSLATE: bytes = _build_mark_table()

#: 4-entry table mapping 2-bit slot values back to ASCII bytes.
_UNPACK_TABLE: bytes = b"ACGT"

#: 256-entry decode lookup: each body byte maps to its 4-byte ACGT
#: expansion (high two bits → first base, low two bits → fourth).
#: Used by :func:`decode` to unpack the body in a single C-driven
#: ``b"".join(...)`` over a list comprehension, which is several
#: times faster than a per-slot Python shift-and-mask loop.
_UNPACK_BYTE_TABLE: tuple[bytes, ...] = tuple(
    bytes(
        (
            _UNPACK_TABLE[(i >> 6) & 0b11],
            _UNPACK_TABLE[(i >> 4) & 0b11],
            _UNPACK_TABLE[(i >> 2) & 0b11],
            _UNPACK_TABLE[i & 0b11],
        )
    )
    for i in range(256)
)


# ── Public API ─────────────────────────────────────────────────────


def encode(data: bytes) -> bytes:
    """Encode ``data`` using BASE_PACK + sidecar mask.

    Returns a self-contained byte string per the wire format
    described in this module's docstring (HANDOFF.md M84 §2). Pure
    ACGT input compresses to ~25% of original size plus a 13-byte
    header; non-ACGT bytes round-trip losslessly via the mask.

    Parameters
    ----------
    data:
        Input bytes. Any byte value is accepted; ``A``/``C``/``G``/``T``
        (uppercase only — see binding decision §81) are packed into
        2-bit slots, anything else gets a mask entry.

    Returns
    -------
    bytes
        Encoded stream of length
        ``13 + ceil(len(data) / 4) + 5 * mask_count``.
    """
    orig_len = len(data)
    packed_len = (orig_len + 3) // 4

    # Step 1: translate every input byte to its 2-bit slot value
    # (non-ACGT bytes map to placeholder 0b00 here; the mask scan
    # below recovers their true bytes). ``bytes.translate`` runs in
    # C and is ~2 orders of magnitude faster than a Python loop.
    slots = data.translate(_PACK_TRANSLATE)

    # Step 2: pack four slot bytes into one body byte (big-endian
    # within byte: first base in the highest two bits). Pad the
    # slot stream to a multiple of 4 with placeholder zeros so the
    # packing loop need not special-case the tail (binding decision
    # §83 — padding bits in the final body byte are zero).
    pad = (-orig_len) & 3
    if pad:
        slots = slots + bytes(pad)
    it = iter(slots)
    body = bytes(
        (a << 6) | (b << 4) | (c << 2) | d for a, b, c, d in zip(it, it, it, it)
    )
    # Sanity: the zip(...) idiom generated exactly ``packed_len`` bytes.
    assert len(body) == packed_len

    # Step 3: collect mask entries. Most realistic genomic input is
    # >99 % ACGT, so we first do a single ``bytes.translate``-based
    # mark + ``in`` membership probe to short-circuit the no-mask
    # case (the dominant fast path).
    marks = data.translate(_MARK_TRANSLATE)
    if b"\x01" in marks:
        # bytearray.find loop is faster than a per-byte enumerate.
        mask_parts: list[bytes] = []
        start = 0
        while True:
            pos = marks.find(b"\x01", start)
            if pos < 0:
                break
            mask_parts.append(struct.pack(">IB", pos, data[pos]))
            start = pos + 1
        mask_count = len(mask_parts)
        header = struct.pack(">BIII", VERSION, orig_len, packed_len, mask_count)
        return header + body + b"".join(mask_parts)

    header = struct.pack(">BIII", VERSION, orig_len, packed_len, 0)
    return header + body


def decode(encoded: bytes) -> bytes:
    """Decode a stream produced by :func:`encode`.

    Reads the header, unpacks the 2-bit body, applies the sidecar
    mask. Validates strictly: the version byte, the
    ``packed_length`` invariant, the total stream length, and that
    every mask position is in ``[0, original_length)`` and strictly
    ascending.

    Raises
    ------
    ValueError
        If the stream is shorter than the header, has a bad version
        byte, has a wrong ``packed_length`` for the declared
        ``original_length``, has a body or mask section of the
        wrong length, has a mask position out of range, or has
        unsorted / duplicate mask positions.
    """
    if len(encoded) < HEADER_LEN:
        raise ValueError(
            f"BASE_PACK stream too short for header: {len(encoded)} < {HEADER_LEN}"
        )

    version, orig_len, packed_len, mask_count = struct.unpack(
        ">BIII", encoded[:HEADER_LEN]
    )

    if version != VERSION:
        raise ValueError(
            f"BASE_PACK bad version byte: 0x{version:02x} (expected 0x{VERSION:02x})"
        )

    expected_packed = (orig_len + 3) // 4
    if packed_len != expected_packed:
        raise ValueError(
            f"BASE_PACK packed_length mismatch: {packed_len} != "
            f"ceil({orig_len}/4) = {expected_packed}"
        )

    expected_total = HEADER_LEN + packed_len + MASK_ENTRY_LEN * mask_count
    if len(encoded) != expected_total:
        raise ValueError(
            f"BASE_PACK stream length mismatch: {len(encoded)} != "
            f"{expected_total} (header {HEADER_LEN} + body {packed_len} + "
            f"mask {MASK_ENTRY_LEN}*{mask_count})"
        )

    body = encoded[HEADER_LEN : HEADER_LEN + packed_len]

    # Unpack body via 256-entry table → list of 4-byte chunks →
    # ``b"".join``. ``operator.itemgetter(*body)(table)`` runs the
    # entire mapping in C; it returns a single bytes object when
    # body has length 1, so we special-case that.
    table = _UNPACK_BYTE_TABLE
    if orig_len == 0:
        out = bytearray(0)
    else:
        full_bytes = orig_len >> 2
        tail = orig_len - (full_bytes << 2)
        if tail == 0:
            if full_bytes == 1:
                out = bytearray(table[body[0]])
            else:
                out = bytearray(b"".join(operator.itemgetter(*body)(table)))
        else:
            # Decode the full bytes via the fast path, then append a
            # truncated expansion of the final padded byte.
            if full_bytes == 0:
                decoded = b""
            elif full_bytes == 1:
                decoded = table[body[0]]
            else:
                decoded = b"".join(
                    operator.itemgetter(*body[:full_bytes])(table)
                )
            tail_full = table[body[full_bytes]]
            out = bytearray(decoded + tail_full[:tail])

    # Apply mask. Validate ascending positions and 0 <= pos < orig_len
    # in a single scan.
    mask_offset = HEADER_LEN + packed_len
    prev_pos = -1
    for k in range(mask_count):
        entry = encoded[mask_offset + k * MASK_ENTRY_LEN : mask_offset + (k + 1) * MASK_ENTRY_LEN]
        pos, byte = struct.unpack(">IB", entry)
        if pos >= orig_len:
            raise ValueError(
                f"BASE_PACK mask position {pos} out of range "
                f"[0, {orig_len})"
            )
        if pos <= prev_pos:
            raise ValueError(
                f"BASE_PACK mask positions not strictly ascending: "
                f"{pos} after {prev_pos}"
            )
        prev_pos = pos
        out[pos] = byte

    return bytes(out)

File: java/src/main/java/global/thalion/ttio/codecs/BasePack.java
/*
 * TTI-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio.codecs;

/**
 * BASE_PACK genomic-sequence codec — 2-bit ACGT + sidecar mask.
 *
 * <p>Clean-room implementation matching the Python reference
 * ({@code python/src/ttio/codecs/base_pack.py}) byte-for-byte.
 * The 2-bit-per-base packing convention is decades-old prior art,
 * fundamental and ungatewayed by IP. <b>No htslib, no jbzip, no CRAM
 * tools-Java source consulted.</b> The sidecar mask layout (sparse
 * position+byte list) is a TTI-O-specific design choice — see
 * HANDOFF.md M84 §3, binding decision §80.
 *
 * <p>Cross-language equivalents:
 * <ul>
 *   <li>Python: {@code ttio.codecs.base_pack}</li>
 *   <li>Objective-C: {@code TTIOBasePack}</li>
 * </ul>
 *
 * <p>Wire format (big-endian throughout, self-contained):
 * <pre>
 *   Offset   Size   Field
 *   ──────   ────   ─────────────────────────────────────────────
 *   0        1      version            (0x00)
 *   1        4      original_length    (uint32 BE)
 *   5        4      packed_length      (uint32 BE — = ceil(orig/4))
 *   9        4      mask_count         (uint32 BE)
 *   13       var    packed_body        (packed_length bytes)
 *   13+pl    var    mask               (mask_count × 5 bytes:
 *                                          uint32 BE position,
 *                                          uint8 original_byte)
 * </pre>
 *
 * <p>Total length = {@code 13 + packed_length + 5 * mask_count} bytes.
 *
 * <p>Pack mapping (case-sensitive; binding decision §81):
 * <pre>
 *   'A' (0x41) → 0b00
 *   'C' (0x43) → 0b01
 *   'G' (0x47) → 0b10
 *   'T' (0x54) → 0b11
 *   anything else → mask entry (placeholder 0b00 in body)
 * </pre>
 *
 * <p>Bit order within byte is <b>big-endian</b> (binding decision §82) —
 * the first base in the input occupies the two highest-order bits.
 * Padding bits in the final body byte (when {@code len(input) % 4 != 0})
 * are zero (binding decision §83); the decoder uses
 * {@code original_length} to know how many slots to consume.
 *
 * <p>Mask entries are sorted ascending by position (binding decision §84);
 * the encoder emits them in input order, the decoder validates strict
 * ascending order. The first byte is {@code version = 0x00} (binding
 * decision §85), distinct from the M79 codec id {@code 0x06}.
 */
public final class BasePack {

    // ── Wire-format constants ───────────────────────────────────────

    /** Version byte — first byte of every BASE_PACK stream. */
    private static final byte VERSION = 0x00;

    /** Header length: 1 (version) + 4 (orig_len) + 4 (packed_len) + 4 (mask_count). */
    private static final int HEADER_LEN = 13;

    /** Mask entry size: uint32 BE position + uint8 original byte. */
    private static final int MASK_ENTRY_LEN = 5;

    // ── Pack lookup table ───────────────────────────────────────────

    /**
     * 256-entry table mapping every input byte to a 2-bit slot value:
     * A→0, C→1, G→2, T→3, every other byte → 0 (placeholder).
     */
    private static final byte[] PACK_TABLE = buildPackTable();

    private static byte[] buildPackTable() {
        byte[] tbl = new byte[256]; // default 0 — placeholder for non-ACGT
        tbl['A'] = 0b00;
        tbl['C'] = 0b01;
        tbl['G'] = 0b10;
        tbl['T'] = 0b11;
        return tbl;
    }

    private BasePack() {
        // Utility class — non-instantiable.
    }

    // ── Public API ──────────────────────────────────────────────────

    /**
     * Encode {@code data} using BASE_PACK + sidecar mask.
     *
     * <p>Returns a self-contained byte array per the wire format
     * described in this class's javadoc. Pure ACGT input compresses
     * to ~25% of original size plus a 13-byte header; non-ACGT bytes
     * round-trip losslessly via the mask.
     *
     * @param data input bytes (may be empty); {@code null} not allowed.
     * @return encoded stream of length
     *         {@code 13 + ceil(len(data)/4) + 5 * mask_count}.
     * @throws IllegalArgumentException if {@code data} is null.
     */
    public static byte[] encode(byte[] data) {
        if (data == null) {
            throw new IllegalArgumentException("BASE_PACK encode: data must not be null");
        }
        int origLen = data.length;
        int packedLen = (origLen + 3) >>> 2;

        // First pass: count mask entries (non-ACGT bytes).
        int maskCount = 0;
        for (int i = 0; i < origLen; i++) {
            int b = Byte.toUnsignedInt(data[i]);
            if (b != 'A' && b != 'C' && b != 'G' && b != 'T') {
                maskCount++;
            }
        }

        int total = HEADER_LEN + packedLen + MASK_ENTRY_LEN * maskCount;
        byte[] out = new byte[total];

        // Header: version + orig_len + packed_len + mask_count.
        out[0] = VERSION;
        writeUInt32BE(out, 1, origLen);
        writeUInt32BE(out, 5, packedLen);
        writeUInt32BE(out, 9, maskCount);

        // Body: pack 4 slot values per byte, big-endian within byte.
        // The pack table maps non-ACGT to placeholder 0b00; the mask
        // section recovers the true bytes during decode.
        int bodyOff = HEADER_LEN;
        int i = 0;
        int fullChunks = origLen >>> 2;
        for (int chunk = 0; chunk < fullChunks; chunk++) {
            int s0 = PACK_TABLE[Byte.toUnsignedInt(data[i    ])] & 0xFF;
            int s1 = PACK_TABLE[Byte.toUnsignedInt(data[i + 1])] & 0xFF;
            int s2 = PACK_TABLE[Byte.toUnsignedInt(data[i + 2])] & 0xFF;
            int s3 = PACK_TABLE[Byte.toUnsignedInt(data[i + 3])] & 0xFF;
            out[bodyOff++] = (byte) ((s0 << 6) | (s1 << 4) | (s2 << 2) | s3);
            i += 4;
        }
        // Tail (1, 2, or 3 leftover bases). Padding slots are zero
        // (binding decision §83).
        int tail = origLen - (fullChunks << 2);
        if (tail > 0) {
            int s0 = PACK_TABLE[Byte.toUnsignedInt(data[i])] & 0xFF;
            int s1 = (tail >= 2) ? (PACK_TABLE[Byte.toUnsignedInt(data[i + 1])] & 0xFF) : 0;
            int s2 = (tail >= 3) ? (PACK_TABLE[Byte.toUnsignedInt(data[i + 2])] & 0xFF) : 0;
            out[bodyOff++] = (byte) ((s0 << 6) | (s1 << 4) | (s2 << 2));
        }

        // Mask: sparse list of (position, original_byte). Natural
        // left-to-right scan emits entries already sorted ascending
        // (binding decision §84).
        int maskOff = HEADER_LEN + packedLen;
        for (int j = 0; j < origLen; j++) {
            int b = Byte.toUnsignedInt(data[j]);
            if (b != 'A' && b != 'C' && b != 'G' && b != 'T') {
                writeUInt32BE(out, maskOff, j);
                out[maskOff + 4] = (byte) b;
                maskOff += MASK_ENTRY_LEN;
            }
        }

        return out;
    }

    /**
     * Decode a stream produced by {@link #encode(byte[])}.
     *
     * <p>Reads the header, unpacks the 2-bit body, applies the sidecar
     * mask. Validates strictly: the version byte, the
     * {@code packed_length} invariant, the total stream length, and
     * that every mask position is in {@code [0, original_length)} and
     * strictly ascending.
     *
     * @param encoded encoded stream.
     * @return decoded bytes, equal to the encoder's input.
     * @throws IllegalArgumentException if the stream is shorter than
     *         the header, has a bad version byte, has a wrong
     *         {@code packed_length} for the declared
     *         {@code original_length}, has a body or mask section of
     *         the wrong length, has a mask position out of range, or
     *         has unsorted/duplicate mask positions.
     */
    public static byte[] decode(byte[] encoded) {
        if (encoded == null) {
            throw new IllegalArgumentException("BASE_PACK decode: input must not be null");
        }
        if (encoded.length < HEADER_LEN) {
            throw new IllegalArgumentException(
                "BASE_PACK stream too short for header: " + encoded.length
                    + " < " + HEADER_LEN);
        }

        int version = Byte.toUnsignedInt(encoded[0]);
        if (version != (VERSION & 0xFF)) {
            throw new IllegalArgumentException(
                "BASE_PACK bad version byte: 0x"
                    + String.format("%02x", version) + " (expected 0x00)");
        }

        long origLenU = readUInt32BE(encoded, 1);
        long packedLenU = readUInt32BE(encoded, 5);
        long maskCountU = readUInt32BE(encoded, 9);
        if (origLenU > Integer.MAX_VALUE) {
            throw new IllegalArgumentException(
                "BASE_PACK declared original_length too large: " + origLenU);
        }
        if (packedLenU > Integer.MAX_VALUE) {
            throw new IllegalArgumentException(
                "BASE_PACK declared packed_length too large: " + packedLenU);
        }
        if (maskCountU > Integer.MAX_VALUE) {
            throw new IllegalArgumentException(
                "BASE_PACK declared mask_count too large: " + maskCountU);
        }
        int origLen = (int) origLenU;
        int packedLen = (int) packedLenU;
        int maskCount = (int) maskCountU;

        int expectedPacked = (origLen + 3) >>> 2;
        if (packedLen != expectedPacked) {
            throw new IllegalArgumentException(
                "BASE_PACK packed_length mismatch: " + packedLen
                    + " != ceil(" + origLen + "/4) = " + expectedPacked);
        }

        long expectedTotal = (long) HEADER_LEN + packedLen
            + (long) MASK_ENTRY_LEN * maskCount;
        if (encoded.length != expectedTotal) {
            throw new IllegalArgumentException(
                "BASE_PACK stream length mismatch: " + encoded.length
                    + " != " + expectedTotal + " (header " + HEADER_LEN
                    + " + body " + packedLen + " + mask "
                    + MASK_ENTRY_LEN + "*" + maskCount + ")");
        }

        // Unpack body. The natural high-to-low extraction reads the
        // first base from the two highest-order bits of each byte.
        byte[] out = new byte[origLen];
        int bodyOff = HEADER_LEN;
        int fullChunks = origLen >>> 2;
        int o = 0;
        for (int chunk = 0; chunk < fullChunks; chunk++) {
            int b = Byte.toUnsignedInt(encoded[bodyOff + chunk]);
            out[o    ] = unpackTable((b >>> 6) & 0b11);
            out[o + 1] = unpackTable((b >>> 4) & 0b11);
            out[o + 2] = unpackTable((b >>> 2) & 0b11);
            out[o + 3] = unpackTable(b & 0b11);
            o += 4;
        }
        int tail = origLen - (fullChunks << 2);
        if (tail > 0) {
            int b = Byte.toUnsignedInt(encoded[bodyOff + fullChunks]);
            out[o] = unpackTable((b >>> 6) & 0b11);
            if (tail >= 2) {
                out[o + 1] = unpackTable((b >>> 4) & 0b11);
            }
            if (tail >= 3) {
                out[o + 2] = unpackTable((b >>> 2) & 0b11);
            }
        }

        // Apply mask. Validate ascending positions and 0 <= pos < orig_len
        // in a single scan (binding decisions §84, §92).
        int maskOff = HEADER_LEN + packedLen;
        long prevPos = -1L;
        for (int k = 0; k < maskCount; k++) {
            int entryOff = maskOff + k * MASK_ENTRY_LEN;
            long pos = readUInt32BE(encoded, entryOff);
            int origByte = Byte.toUnsignedInt(encoded[entryOff + 4]);
            if (pos >= origLen) {
                throw new IllegalArgumentException(
                    "BASE_PACK mask position " + pos + " out of range [0, "
                        + origLen + ")");
            }
            if (pos <= prevPos) {
                throw new IllegalArgumentException(
                    "BASE_PACK mask positions not strictly ascending: "
                        + pos + " after " + prevPos);
            }
            prevPos = pos;
            out[(int) pos] = (byte) origByte;
        }

        return out;
    }

    // ── Internal helpers ────────────────────────────────────────────

    /** Map a 2-bit slot value (0..3) back to the ASCII byte 'A'/'C'/'G'/'T'. */
    private static byte unpackTable(int slot) {
        // 0 → 'A' (0x41), 1 → 'C' (0x43), 2 → 'G' (0x47), 3 → 'T' (0x54).
        switch (slot) {
            case 0:  return (byte) 'A';
            case 1:  return (byte) 'C';
            case 2:  return (byte) 'G';
            default: return (byte) 'T';
        }
    }

    private static void writeUInt32BE(byte[] buf, int off, int val) {
        buf[off]     = (byte) ((val >>> 24) & 0xFF);
        buf[off + 1] = (byte) ((val >>> 16) & 0xFF);
        buf[off + 2] = (byte) ((val >>> 8) & 0xFF);
        buf[off + 3] = (byte) (val & 0xFF);
    }

    private static long readUInt32BE(byte[] buf, int off) {
        return ((long) Byte.toUnsignedInt(buf[off])     << 24)
             | ((long) Byte.toUnsignedInt(buf[off + 1]) << 16)
             | ((long) Byte.toUnsignedInt(buf[off + 2]) << 8)
             | ((long) Byte.toUnsignedInt(buf[off + 3]));
    }
}
